src/cargo/core/compiler/fingerprint/mod.rs

```diff
@@ -1860,11 +1860,18 @@ where
     for path in paths {
         let path = path.as_ref();
 
-        // Assuming anything in cargo_home is immutable (see also #9455 about marking it readonly)
-        // which avoids rebuilds when CI caches $CARGO_HOME/registry/{index, cache} and
-        // $CARGO_HOME/git/db across runs, keeping the content the same but changing the mtime.
-        if let Ok(true) = home::cargo_home().map(|home| path.starts_with(home)) {
-            continue;
+        // Assuming anything in cargo_home/{git, registry} is immutable
+        // (see also #9455 about marking cargo_home readonly) which avoids rebuilds when CI caches
+        // $CARGO_HOME/registry/{index, cache} and $CARGO_HOME/git/db across runs, keeping the
+        // content the same but changing the mtime.
+        if let Ok(cargo_home) = home::cargo_home() {
+            let is_readonly = ["git", "registry"]
+                .into_iter()
+                .map(|subfolder| cargo_home.join(subfolder))
+                .any(|readonly_dir| path.starts_with(readonly_dir));
```

Comment:
* Maybe just a micro optimization. This `cargo_home.join` is in a hot loop over `paths`. Could we do the allocation out of the loop upfront?
* I am a big wary that we now duplicate the path construction logic. Do you feel it better if we reuse the code [here in `Config`](https://github.com/rust-lang/cargo/blob/20df9e40a4d41dd08478549915588395e55efb4c/src/cargo/util/config/mod.rs#L353-L361)?

Reply:
I will take a look at it. Should I combine the commits, when I'm done?

Reply:
My personal preference is making each commit atomic. Given the size of this change is not huge, it makes sense to me just rebase and leave one or two commits.

More personally, I may have the first commit to have a test first asserting the "bug behavior". Then the second contains the actual patch and the fixup of the test.